app/util/links.ts

```diff
@@ -71,3 +73,117 @@ export function getStatusLink(urlRoot: string, jobID: string, rel: string): Link
     type: 'application/json',
   };
 }
+
+/**
+ * Given a job event, return the JobLink representing the corresponding event.
+ * @param event - the event being triggered by the JobLink
+ * @param jobID - the jobID that this link is for
+ * @param urlRoot - the root url to use in the link
+ * @param isAdmin - whether to include /admin in the link
+ * @returns - the job link for the job event
+ */
+function getLinkForJobEvent(
+  event: JobEvent, 
+  jobID: string, 
+  urlRoot: string, 
+  isAdmin = false,
+): JobLink {
+  const adminPath = isAdmin ? '/admin' : '';
+  switch (event) {
+    case JobEvent.RESUME:
+      return new JobLink({
+        title: 'Resumes the job.',
+        href: `${urlRoot + adminPath}/jobs/${jobID}/resume`,
+        type: 'application/json',
+        rel: 'resumer',
+      });
+    case JobEvent.SKIP_PREVIEW:
+      return new JobLink({
+        title: 'Skips preview and runs the job.',
+        href: `${urlRoot + adminPath}/jobs/${jobID}/skip-preview`,
+        type: 'application/json',
+        rel: 'preview-skipper',
+      });
+    case JobEvent.CANCEL:
+      return new JobLink({
+        title: 'Cancels the job.',
+        href: `${urlRoot + adminPath}/jobs/${jobID}/cancel`,
+        type: 'application/json',
+        rel: 'canceler',
+      });
+    case JobEvent.PAUSE:
+      return new JobLink({
+        title: 'Pauses the job.',
+        href: `${urlRoot + adminPath}/jobs/${jobID}/pause`,
+        type: 'application/json',
+        rel: 'pauser',
+      });
+  }
+}
+
+/**
+ * Return a set of JobEvents representing the actions that are currently available
+ * to a user for a particular job.
+ * Note that this only returns actions that users can precipitate via state change links
+ * (e.g. JobEvent.FAIL will not be returned).
+ * @param job - the job to return valid actions (JobEvents) for
+ * @returns a set of JobEvent
+ */
+export function getLinkRelevantJobEvents(job: Job): Set<JobEvent> {
+  const transitions: [JobEvent, JobStatus][] = [
+    // [event, resultant status]
+    [JobEvent.CANCEL, JobStatus.CANCELED],
```

Comment:
I don't like how we are somewhat leaking logic out of the state machine and into the code here. I would like it better if we could leverage the state machine definition to simply check all state/event/state transitions here to determine what is viable. I realize that there are some events that don't make sense for a user and aren't captured by the state machine (yet), so you don't need to change this, but we should give this some thought down the line.

Reply:
I think I get what you're saying. The job.ts class (pause method for example) `validateTransition(this.status, JobStatus.PAUSED, JobEvent.PAUSE)` seems to be doing the same type of thing that we're doing here, so is that leaking as well? Or is the fact that this logic is not staying in the job.ts class causing the "leak"? I could easily move this list of transitions and this method for that matter back into that job class if that helps. Also, not sure if it's relevant but maybe this could be useful for encapsulating more logic in the state machine https://xstate.js.org/docs/guides/guards.html#guards-condition-functions

Even better, would the goal be to ideally move _all_ of this to the state machine (rather than simply the job class)? Just trying to make sure I understand.

Reply:
OK, sounds good.